### ISSUE TITLE
List Continuation for Markdown

### ISSUE THREAD

**Issue by benjamin-kirkbride** [Akuli/porcupine#1305]
I would like to see automatic list continuation when editing markdown, as well as some other features like that.

Is such a thing able to be done with a plugin? Any tips?

Thanks :)

**Comment by benjamin-kirkbride** [comment]
Example of list continuation:
type `- first item` followed by `ENTER` should result in:
```markdown
- first item
- |
```
where `|` is the cursor

**Comment by rdbende** [comment]
It could be implemented in a plugin similarly to how the autoindent works.

Some IDEs also auto-close quotes and parentheses. Porcupine doesn't do this, though it would also be a useful feature. It might would be hard to get right.

**Comment by benjamin-kirkbride** [comment]
Yeah I was just looking at https://github.com/Akuli/porcupine/blob/main/porcupine/plugins/autoindent.py

**Comment by Moosems** [comment]
@rdbende it was easily implemented in DIP...

**Comment by benjamin-kirkbride** [comment]
@Moosems what do you mean?

**Comment by Moosems** [comment]
Just make a dict of items to insert certain variations on and then context for each variation.

For example:
```json
{
    "python":
    "indent regex":
        [
            "action"
        ]
    ...
}
```

**Comment by Moosems** [comment]
Or even just a bind and associated action.

**Comment by benjamin-kirkbride** [comment]
Is there some prior art here? Sorry I'm starting from ground zero here

**Comment by rdbende** [comment]
> Just make a dict of items to insert certain variations on and then context for each variation.

For example:

{
    "python":
    {
        "indent regex":
            [
                "action"
            ]
        ...
    }
}

Umm, wat??

**Comment by rdbende** [comment]
@benjamin-kirkbride So I'm guessing @Moosems is trying to say, that you could put some configuration about these auto-continuations into the `filetype.toml` file.
It could be a regular expression that you define similarly to how the autoindent regexes are. Then you can use this regex to find out when it should continue the markdown list, or even to auto-close a quote in some other language. This part could be quite similar to the autoindent plugin.

**Comment by benjamin-kirkbride** [comment]
Ah, I see.

Any thoughts on `TAB` moving the list up, such that when in this position:
```markdown
- item 1
- |
```

pressing `TAB` would result in:
```markdown
- item 1
    - |
```
?

**Comment by Akuli** [comment]
I like both of your ideas :) Ideally they would be in the same plugin. For the initial version, the plugin could just check whether the filetype is set to Markdown. We can later make it configurable in `filetypes.toml`, so that other similar formats (e.g. RST) can benefit from it too.

> @Moosems what do you mean?

DIP is Moosems's closed-source editor project. Maybe the "For example" thing was part of a configuration file for DIP.

You don't really need to know anything about DIP to work on Porcupine, but you will occasionally see Moosems talking about it on Porcupine's issue tracker.

**Comment by Moosems** [comment]
Yes you will 🤣. In DIP (my personal project) I use a list of chars to finish with and their endings. This could be updated to check a line for its contents by regex and cases but that was the idea I presented :).
